core-bundle/src/Twig/Inheritance/DynamicUseTokenParser.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Inheritance;
+
+use Contao\CoreBundle\Twig\ContaoTwigUtil;
+use Twig\Error\SyntaxError;
+use Twig\Node\Expression\ConstantExpression;
+use Twig\Node\Node;
+use Twig\Token;
+use Twig\TokenParser\AbstractTokenParser;
+
+/**
+ * This parser is a drop in replacement for @see \Twig\TokenParser\UseTokenParser
+ * that adds support for the Contao template hierarchy.
+ *
+ * @experimental
+ */
+final class DynamicUseTokenParser extends AbstractTokenParser
+{
+    public function __construct(private TemplateHierarchyInterface $hierarchy)
+    {
+    }
+
+    public function parse(Token $token): Node
+    {
+        $template = $this->parser->getExpressionParser()->parseExpression();
+        $stream = $this->parser->getStream();
+
+        if (!$template instanceof ConstantExpression) {
+            throw new SyntaxError('The template references in a "use" statement must be a string.', $stream->getCurrent()->getLine(), $stream->getSourceContext());
+        }
+
+        $this->adjustTemplateName($stream->getSourceContext()->getPath(), $template);
```

Comment:
The `getSourceContext()` method is marked `@internal`. Is there a way for us to work around this?

Reply:
No, unfortunately (same for the `extends` tag btw.).

It is very unlikely that this functionality will be removed, though, because Twig needs it itself. But we should probably pitch a change in the Twig repo to make it a public API once our implementation has settled.